example_client/cpp/src/resnet_client_benchmark.cpp

```diff
@@ -0,0 +1,452 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <thread>
+
+#include "google/protobuf/map.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/command_line_flags.h"
+#include "tensorflow_serving/apis/prediction_service.grpc.pb.h"
+
+#include "common.hpp"
+#include "grpcpp/create_channel.h"
+#include "grpcpp/security/credentials.h"
+#include "opencv2/opencv.hpp"
+
+using grpc::Channel;
+using grpc::ClientAsyncResponseReader;
+using grpc::ClientContext;
+using grpc::CompletionQueue;
+using grpc::Status;
+
+using tensorflow::serving::PredictionService;
+using tensorflow::serving::PredictRequest;
+using tensorflow::serving::PredictResponse;
+
+template <typename T>
+struct AsyncClientCall {
+    PredictResponse reply;
+    ClientContext context;
+    Status status;
+    std::unique_ptr<ClientAsyncResponseReader<PredictResponse>> response_reader;
+    std::vector<T> selectedEntries;
+    tensorflow::int64 id;
+};
+
+struct Configuration {
+    tensorflow::string address = "localhost";
+    tensorflow::string port = "9000";
+    tensorflow::string modelName = "resnet";
+    tensorflow::string inputName = "0";
+    tensorflow::string outputName = "1463";
+    tensorflow::int64 iterations = 10;
+    tensorflow::int64 batchSize = 1;
+    tensorflow::string imagesListPath = "input_images.txt";
+    tensorflow::string layout = "nchw";
+    tensorflow::int64 producers = 1;
+    tensorflow::int64 consumers = 8;
+    tensorflow::int64 max_parallel_requests = 100;
+    tensorflow::int64 benchmark_mode = 0;
+
+    bool validate() const {
+        if (imagesListPath.empty())
+            return false;
+        if (batchSize < 0)
+            return false;
+        if (iterations < 0)
+            return false;
```

Comment:
Is accepting 0 safe?

Reply:
0 means taking all entries from input_images.txt

Reply:
So `iterations` == 0 means that we run all available images once?
What about `batchSize`? Do we load all available images to a single batch for `batchSize` == 0?

Reply:
I disallowed passing 0 values, there is not big value in accepting it